testnet/stacks-node/src/nakamoto_node/relayer.rs

```diff
@@ -378,89 +399,218 @@ impl RelayerThread {
+                // If so, then we can issue a BlockFound.
+                SortitionDB::get_ancestor_snapshot(
+                    &ih,
+                    stacks_tip_sn.block_height,
+                    &last_winning_snapshot.sortition_id,
+                )
+                .map_err(|e| {
+                    error!("Relayer: Failed to load ancestor snapshot: {e:?}");
+                    e
+                })
+                .ok()
+                .flatten()
+                .map(|sn| {
+                    let need_blockfound = sn.consensus_hash == stacks_tip_sn.consensus_hash;
+                    if !need_blockfound {
+                        test_debug!(
+                            "Relayer: stacks_tip_sn.consensus_hash ({}) != sn.consensus_hash ({})",
+                            &stacks_tip_sn.consensus_hash,
+                            &sn.consensus_hash
+                        );
+                    }
+                    need_blockfound
+                })
+                .unwrap_or_else(|| {
+                    test_debug!(
+                        "Relayer: no ancestor at height {} off of sortition {} height {}",
+                        stacks_tip_sn.block_height,
+                        &last_winning_snapshot.consensus_hash,
+                        last_winning_snapshot.block_height
+                    );
+                    false
                 })
+            };
+            if need_blockfound {
+                info!(
+                    "Relayer: will submit late BlockFound for {}",
+                    &last_winning_snapshot.consensus_hash
+                );
+                // prepare to extend after our BlockFound gets mined.
+                self.tenure_extend_timeout = Some(Instant::now());
+                return Some(MinerDirective::BeginTenure {
+                    parent_tenure_start: StacksBlockId(
+                        last_winning_snapshot.winning_stacks_block_hash.clone().0,
+                    ),
+                    burnchain_tip: last_winning_snapshot,
+                    late: true,
+                });
             }
-        };
-        directive
+        }
+
+        // try to continue our tenure if we produced the canonical Stacks tip.
+        if stacks_tip_sn.miner_pk_hash == self.get_mining_key_pkh() {
+            info!("Relayer: No sortition, but we produced the canonical Stacks tip. Will continue tenure.");
+
+            if last_winning_snapshot.miner_pk_hash != self.get_mining_key_pkh() {
+                // delay trying to continue since the last snasphot with a sortition was won
+                // by someone else -- there's a chance that this other miner will produce a
+                // BlockFound in the interim.
+                debug!("Relayer: Did not win last winning snapshot despite mining the ongoing tenure, so allowing the new miner some time to come online.");
+                self.tenure_extend_timeout = Some(Instant::now());
```

Comment:
Should this be `Instant::now()`, or `Instant::now() + config.tenure_extend_timeout`?